subsys/random/Kconfig

```diff
@@ -6,18 +6,32 @@
 menu "Random Number Generators"
 
 config TEST_RANDOM_GENERATOR
-	bool "Non-random number generator"
-	depends on !ENTROPY_HAS_DRIVER
+	bool "Allow non-random number generator"
 	help
-	  This option signifies that the kernel's random number APIs are
-	  permitted to return values that are not truly random.
-	  This capability is provided for testing purposes, when a truly random
+	  This option signifies that a non-random number generator is allowed to
+	  be used and the kernel's random number APIs are permitted to return
+	  values that are not truly random.
+
+	  This capability is provided for testing purposes when a truly random
 	  number generator is not available. The non-random number generator
 	  should not be used in a production environment.
 
+	  This option is intended to be selected only by application-level
+	  configurations (e.g. in tests and samples) to indicate that the
+	  application is allowed to run with a random number generator that is not
+	  truly random. Board-level configurations must not select this option
+	  unless the sole purpose of the board is testing (e.g. QEMU emulation
+	  boards).
+
+	  Note that this option does not imply that a non-random number generator
+	  is selected -- that is indicated by RNG_GENERATOR_CHOICE. An entropy
+	  device-backed random number generator, if available, will be selected by
+	  default even when CONFIG_TEST_RANDOM_GENERATOR=y.
+
 choice RNG_GENERATOR_CHOICE
 	prompt "Random generator"
-	default ENTROPY_DEVICE_RANDOM_GENERATOR
+	default ENTROPY_DEVICE_RANDOM_GENERATOR if ENTROPY_HAS_DRIVER
+	default TIMER_RANDOM_GENERATOR if TEST_RANDOM_GENERATOR
```

Comment:
The fallback to utilize a timer peripheral as a "pseudo entropy source" has security implications in your system. There exists as-of-now very little protection against this ending up in an end-product due to misconfigured DTS or driver configuration. 

I have argued in a related feature request that there is an option available to us that goes beyond just printing warnings once TEST_RANDOM_ENTROPY/TIMER_RANDOM_GENERATOR is enabled in the system 

Please see: https://github.com/zephyrproject-rtos/zephyr/issues/53019#issuecomment-1350956828

Reply:
> There exists as-of-now very little protection against this ending up in an end-product due to misconfigured DTS or driver configuration.

IMHO, a very intrusive warning that is printed every time you run CMake is quite sufficient.

> I have argued in a related feature request that there is an option available to us that goes beyond just printing warnings once TEST_RANDOM_ENTROPY/TIMER_RANDOM_GENERATOR is enabled in the system

That is well beyond the scope of the enhancement issue or this PR. That will require a much broader discussion on how to organise and handle different build profiles (preferably, with a detaild RFC on what that would look like).

Reply:
There is a reference to `hardenconfig` which already exists that provides exactly what I wish for (see PR comment) https://github.com/zephyrproject-rtos/zephyr/issues/53019#issuecomment-1350956828) which is showing that TEST_RANDOM_GENERATOR is disallowed when you build with hardenconfig. I would argue that an additional line should be made in this file to also disqualify TIMER_RANDOM_GENERATOR. This could be done in this PR...

https://github.com/zephyrproject-rtos/zephyr/blob/95717520bf4309c7e54fdf6d99975491880fb874/scripts/kconfig/hardened.csv#LL27-L27C22

Please add `TIMER_RANDOM_GENERATOR,n` then my comment on fallback for RNG_GENERATOR_CHOICE is sufficiently addressed

Reply:
> I propose adding `TIMER_RANDOM_GENERATOR,n` to the hardenconfig.

Done.